zlib-rs/src/inflate.rs

```diff
@@ -1054,84 +1383,15 @@ impl State<'_> {
-                    // space available. This means for many SIMD operations we don't need to process a
-                    // remainder; we just copy blindly, and a later operation will overwrite the extra copied
-                    // bytes
-                    if avail_in >= INFLATE_FAST_MIN_HAVE && avail_out >= INFLATE_FAST_MIN_LEFT {
-                        inflate_fast_help(self, 0);
-                        continue 'label;
-                    }
-
-                    self.back = 0;
-
-                    // get a literal, length, or end-of-block code
-                    let mut here;
-                    loop {
-                        let bits = self.bit_reader.bits(self.len_table.bits);
-                        here = self.len_table_get(bits as usize);
-
-                        if here.bits <= self.bit_reader.bits_in_buffer() {
-                            break;
-                        }
-
-                        pull_byte!(self);
-                    }
-
-                    if here.op != 0 && here.op & 0xf0 == 0 {
-                        let last = here;
-                        loop {
-                            let bits = self.bit_reader.bits((last.bits + last.op) as usize) as u16;
-                            here = self.len_table_get((last.val + (bits >> last.bits)) as usize);
-                            if last.bits + here.bits <= self.bit_reader.bits_in_buffer() {
-                                break;
-                            }
-
-                            pull_byte!(self);
-                        }
-
-                        self.bit_reader.drop_bits(last.bits);
-                        self.back += last.bits as usize;
-                    }
-
-                    self.bit_reader.drop_bits(here.bits);
-                    self.back += here.bits as usize;
-                    self.length = here.val as usize;
-
-                    if here.op == 0 {
-                        self.mode = Mode::Lit;
-
-                        continue 'label;
-                    } else if here.op & 32 != 0 {
-                        // end of block
-
-                        // eprintln!("inflate:         end of block");
-
-                        self.back = usize::MAX;
-                        self.mode = Mode::Type;
-
-                        continue 'label;
-                    } else if here.op & 64 != 0 {
-                        self.mode = Mode::Bad;
-
-                        break 'label self.bad("invalid literal/length code\0");
-                    } else {
-                        // length code
-                        self.extra = (here.op & MAX_BITS) as usize;
-                        self.mode = Mode::LenExt;
-
-                        continue 'label;
-                    }
-                }
                 Mode::Len_ => {
                     self.mode = Mode::Len;
 
                     continue 'label;
                 }
+                Mode::Len => match self.len_and_friends() {
```

Comment:
Don't you need a `restore!()` before this call and a `load!()` after it? `len_and_friends` calls `load!()` as one of the first things it does, which would result in an out-of-date load if `restore!()` is missing here.